core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java

```diff
@@ -2871,6 +2876,31 @@ public void testRegisterExistingTable() {
     assertThat(catalog.dropTable(identifier)).isTrue();
   }
 
+  @Test
+  public void testRegisterAndOverwriteExistingTable() {
+    Assumptions.assumeTrue(supportsRegisterTableWithOverwrite());
+    C catalog = catalog();
+
+    TableIdentifier identifier = TableIdentifier.of("a", "t1");
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(identifier.namespace());
+    }
+
+    catalog.createTable(identifier, SCHEMA);
+    Table table = catalog.loadTable(identifier);
+    TableOperations ops = ((BaseTable) table).operations();
+    assertThat(table.spec().isPartitioned()).isFalse();
+    // update table spec
+    table.updateSpec().addField(bucket("id", 16)).commit();
+    String newMetadataLocation = ops.refresh().metadataFileLocation();
+    // register and overwrite
+    catalog.registerTable(identifier, newMetadataLocation, true);
```

Comment:
Just tried out and apparently this test passes without the successful registerTable() call. Probably this goes for other similar tests too in this PR.